libs/base/Control/Monad/Partial.idr

```diff
@@ -0,0 +1,52 @@
+module Control.Monad.Partial
+
+%default total
+
+
+||| Delay/partiality monad. Represents partial computations which may not
+||| terminate or cover all inputs. This can be preferable to defining a partial
+||| Idris function as it allows termination to be proven as an external proof.
+public export
+data Partial : Type -> Type where
+    Now : a -> Partial a
+    Later : Inf (Partial a) -> Partial a
+
+
+public export
+Functor Partial where
+    map f (Now x) = Now (f x)
+    map f (Later x) = Later (map f x)
+
+
+public export
+Applicative Partial where
+    pure = Now
+
+    Now f <*> x = map f x
+    Later f <*> x = Later (f <*> x)
+
+
+public export
+Monad Partial where
+    Now x >>= f = f x
+    Later x >>= f = Later (x >>= f)
+
+
+||| A partial computation which never terminates (aka. undefined/bottom).
+public export
+never : Partial a
+never = Later never
+
+
+public export
+data Total : Partial a -> Type where
+    TNow : Total (Now x)
+    TLater : Total (Force x) -> Total (Later x)
+
+
+||| Extract the value from a partial computation if you have a proof it is
+||| actually total.
+public export
+runPartial : (x : Partial a) -> Total x -> a
```

Comment:
This proof argument can be an `auto`-implicit argument with quantity `0`, this simplifies function implementation, its usage, and actual runtime.

Reply:
I agree it should be erased, but I'm not sure that making it an auto-implicit is the right choice as proving totality will often require case splitting on arguments (which the proof search won't do automatically I think).

Reply:
Are you talking about implementation or calling to this function? I checked, and I agree that you'll need to pass the argument manually, but you can do this easily, say:

```idris
runPartial : (x : Partial a) -> (0 _ : Total x) => a
runPartial (Now v) = v
runPartial (Later x) @{TLater xx} = runPartial x @{xx}
```

But, if you consider this ugly -- then ok, leave it explicit. Anyway, I think, that callers of this function would like to pass this argument automatically -- do you agree?

Reply:
I think that callers would still like to pass it explicitly most of the time, as most of the time you are proving totality for a function rather than a concrete value.

Here is an example of how I imagine this type being used. The function isn't actually useful but is the simplest example I could come up with:

```idris
toZeroP : Nat -> Partial Nat
toZeroP n = if n == 0 then pure n else Later $ toZeroP (pred n)

totality : (n : Nat) -> Total (toZeroP n)
totality Z = TNow
totality (S n) = TLater (totality n)

toZeroT : Nat -> Nat
toZeroT n = runPartial (toZeroP n) (totality n)
```

If we use an auto implicit then it will fail to find a solution as the totality proof is by induction.

Reply:
I would probably write `runPartial : {x : Partial a} -> (0 _ : Total x) -> a` instead tbh. `x` is uniquely determined from `Total x`. :)